### ISSUE TITLE
OAuth2 Token Introspection Authentication

### ISSUE THREAD

**Issue by eglia** [goauthentik/authentik#1234]
**Describe the bug**
So first of all, I'm not sure if it is a bug 😄
I'm trying to get roundcube & dovecot to authenticate using OAuth2 against authentik. I got it working so far that roundcube gets a token and passes it to dovecot. Dovecot then attempts to verify it using the introspection endpoint, but Authentik reports that the token is not valid. I looked through the code and it looks like authentik expects there to be some authentication on the POST request, which dovecot does not do (I think it only has the token from dovecot and nothing else).

**To Reproduce**

**Expected behavior**

**Screenshots**

**Logs**
Authentik log
```
{"event": "Not authenticated", "host": "authentik.io", "level": "debug", "logger": "authentik.providers.oauth2.views.introspection", "pid": 30, "request_id": "aa1366517a2441768591b773441cf071", "timestamp": "2021-08-07T14:20:43.631680"}
{"event": "/application/o/introspect/", "host": "172.20.0.34", "level": "info", "logger": "authentik.asgi", "method": "POST", "pid": 30, "request_id": "aa1366517a2441768591b773441cf071", "runtime": 17, "scheme": "https", "size": 0.024, "status": 200, "timestamp": "2021-08-07T14:20:43.634143"}
```

**Version and Deployment (please complete the following information):**
 - authentik version: 2021.7.2
 - Deployment: docker-compose

**Additional context**
The frame sent by dovecot:
```
POST /application/o/introspect/ HTTP/1.1
Host: authentik.io
Date: Sat, 07 Aug 2021 14:20:43 GMT
User-Agent: dovecot-oauth2-passdb/2.3.4.1
Content-Length: 1287
Connection: Keep-Alive
Content-Type: application/x-www-form-urlencoded

token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjkzN2NhMWIyMzY3Yzk4YTA0NDVlM2RlN2I3MTAyZDQ3In0.eyJpc3MiOiJodHRwczovL2F1dGhlbnRpay5raGFyYW4uY2gvYXBwbGljYXRpb24vby9tYWlsLyIsInN1YiI6IjJiYzkwM2M4OWI4NmQ4NDllMzllMGI1OWY2YjExYjM4ZWQ1ZjU5YTA0NGM1NWNjNTIzNWRmNTI1MDg2NWEyNTMiLCJhdWQiOiI2NzllYmQxYWE0MjZjNTBhZDEzNjU1NGM3MjQwZWVlNzNjYjM0MTUzIiwiZXhwIjoxNjI4MzQ2NjM5LCJpYXQiOjE2MjgzNDYwMzksImF1dGhfdGltZSI6MTYyODIyOTA2MiwiYWNyIjoiZ29hdXRoZW50aWsuaW8vcHJvdmlkZXJzL29hdXRoMi9kZWZhdWx0IiwiY19oYXNoIjpudWxsLCJub25jZSI6bnVsbCwiYXRfaGFzaCI6bnVsbCwiZW1haWwiOiJhbmRyZWFzQGVnbGltYWlsLm5ldCIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJuYW1lIjoiQW5kcmVhcyBFZ2xpIiwiZ2l2ZW5fbmFtZSI6IkFuZHJlYXMgRWdsaSIsImZhbWlseV9uYW1lIjoiIiwicHJlZmVycmVkX3VzZXJuYW1lIjoiYW5kcmVhcyIsIm5pY2tuYW1lIjoiYW5kcmVhcyIsImdyb3VwcyI6WyJBZG1pbnMiLCJVc2VycyIsIkxEQVAgQmluZCBVc2VycyJdLCJjaWQiOiI2NzllYmQxYWE0MjZjNTBhZDEzNjU1NGM3MjQwZWVlNzNjYjM0MTUzIiwidWlkIjoiNDRiZGM2MTFlZmQ0NDlkZTg2YWUzOTYxMjdlOWM3MjgifQ.wmHDhRtQbHI61SdKkCvj5fJ4tCIIONXIq0BS_un9ev4oeh6v4z5ceSsg9yggImDY2kZjoZdyviZYOKohYf9f0kcAbGtD9zoEmBbS2Cf0xhoELdPpnrv2pIsIsdZ7TmIK0w4dwggMoTNOejnS2oGJ1YBcbH4ClSfM179AXLmQgor4aR5Wo_AF5d8xWSaKsI7mBly2ZDVFeGaiygJdvRrp-WBDi0Bb3Dk_H8Df1afXIax1cMGSQTsMcifKnVHFGz7oN3kULzROmFGEGMSs-dxMUfjZ-CWR34gXMtceMxzKaMWiFpczfHyrvNsowIW_uCKuJW2dEhbFjzpuF_BxDo-ks
```
And the response from authentik:
```
HTTP/1.1 200 OK
Server: openresty
Date: Sat, 07 Aug 2021 14:20:43 GMT
Content-Type: application/json
Content-Length: 17
Connection: keep-alive
Cache-Control: no-store
Pragma: no-cache
Referrer-Policy: same-origin
Vary: Accept-Encoding
Vary: Cookie
X-Authentik-Id: aa1366517a2441768591b773441cf071
Strict-Transport-Security: max-age=63072000;includeSubDomains; preload
X-Served-By: authentik.io

{"active": false}
```

*Unrelated:*
I'm also trying to use LDAP authentication in dovecot/postfix with authentik. With the latest fixes from the SSSD issue, this works fine now, except that it is super slow (multiple seconds).

**Comment by BeryJu** [comment]
According to the RFC for OAuth2 Token introspection, authentication is required: https://datatracker.ietf.org/doc/html/rfc7662#section-2.1:

```
   To prevent token scanning attacks, the endpoint MUST also require
   some form of authorization to access this endpoint, such as client
   authentication as described in OAuth 2.0 [RFC6749] or a separate
   OAuth 2.0 access token such as the bearer token described in OAuth
   2.0 Bearer Token Usage [RFC6750].  The methods of managing and
   validating these authentication credentials are out of scope of this
   specification.
```

If you turn the authentik Log-level to debug you should also see whats happening, it'll probably say something like `Not authenticated`.

**Comment by BeryJu** [comment]
https://doc.dovecot.org/configuration_manual/authentication/oauth2/#full-config-file So dovecot can send the Authorization header when using GET, however GET is not a valid method for the Introspection endpoint according to the RFC.

**Comment by eglia** [comment]
Yes, that's the reference I'm using.
I tried adding authentication by setting the url to `client_id:client_secret@authentik.io/application/o/introspect`. This gets me the expected output from authentik, but dovecot expects there to be an "email" field, which does not exist (and should not exist, according to the RFC). I also tried using the userinfo endpoint as tokeninfo_url or introspection_url (which gets me the expected fields from authentik, including email), but then dovecot crashes with a failed assertion in json-parser.c (the data returned by authentik is valid json).
I'm not really sure how dovecot expects this to work, there is little documentation. But it looks to me that authentik behaves as specified by the oauth2 standard.

**Comment by BeryJu** [comment]
I opened this PR to fix the authentication issue: https://github.com/dovecot/core/pull/168.

The name of the email field you can adjust with `username_attribute` if I understand correctly, but authentik doesn't return any relevant value in the introspection endpoint.

For the userinfo endpoint, setting `?access_token` like dovecot does is supported, but I don't know why it would fail with a JSON error.

**Comment by eglia** [comment]
So I got it working.
Dovecot doesn't like the "groups" attribute of the userinfo endpoint. When I remove the it from the authentic openid profile mapping, the authentication works with these settings:
```
tokeninfo_url = https://authentik.io/application/o/userinfo/?access_token=
introspection_url = https://client_id:client_secret@authentik.io/application/o/introspect/
introspection_mode = post
force_introspection = yes
active_attribute = active
active_value = true
tls_ca_cert_dir = /etc/ssl/certs
```
The problem is probably that the groups attribute is an array.

**Comment by xpufx** [comment]
@eglia Can you please post your working configuration for Roundcube?

**Comment by eglia** [comment]
```
# cat oauth2.inc.php
<?php

//$config['per_user_logging'] = false;
//$config['smtp_log'] = true;
//$config['log_logins'] = true;
//$config['session_debug'] = true;
//$config['sql_debug'] = true;
//$config['imap_debug'] = true;
//$config['ldap_debug'] = true;
//$config['smtp_debug'] = true;
//$config['memcache_debug'] = true;
//$config['apc_debug'] = true;
//$config['redis_debug'] = true;

$config['use_https'] = true;
$config['assets_path'] = '/';

$config['login_password_maxlen'] = 4096;

// ----------------------------------
// OAuth
// ----------------------------------

// Enable OAuth2 by defining a provider. Use 'generic' here
$config['oauth_provider'] = 'generic';

// Provider name to be displayed on the login buttin
$config['oauth_provider_name'] = 'Authentik';

// Mandatory: OAuth client ID for your Roundcube installation
$config['oauth_client_id'] = 'CHANGEME';

// Mandatory: OAuth client secret
$config['oauth_client_secret'] = 'CHANGEME'

// Mandatory: URI for OAuth user authentication (redirect)
$config['oauth_auth_uri'] = 'https://authentik.io/application/o/authorize/';

// Mandatory: Endpoint for OAuth authentication requests (server-to-server)
$config['oauth_token_uri'] = 'https://authentik.io/application/o/token/';

// Optional: Endpoint to query user identity if not provided in auth response
$config['oauth_identity_uri'] = 'https://authentik.io/application/o/userinfo/';

// Optional: disable SSL certificate check on HTTP requests to OAuth server
// See http://docs.guzzlephp.org/en/stable/request-options.html#verify for possible values
$config['oauth_verify_peer'] = true;

// Mandatory: OAuth scopes to request (space-separated string)
$config['oauth_scope'] = 'email openid profile';

// Optional: additional query parameters to send with login request (hash array)
$config['oauth_auth_parameters'] = [];

// Optional: array of field names used to resolve the username within the identity information
//$config['oauth_identity_fields'] = ['mailbox'];
$config['oauth_identity_fields'] = ['email'];

// Boolean: automatically redirect to OAuth login when opening Roundcube without a valid session
$config['oauth_login_redirect'] = true;

$prefs['_GLOBAL']['pwstore_scheme'] = 'des_key';
```
Fill in the client ID and client secret and change authentik.io to your domain. Also note my last comment, I had to create a custom openID profile mapping in authentik which does not include the groups of the user.

**Comment by xpufx** [comment]
> Also note my last comment, I had to create a custom openID profile mapping in authentik which does not include the groups of the user.

I have the same exact configuration (for my authentik url) now. It's got to be the group thing then. 

I am posting the error for future readers:
`
errors: <9c5eb26a> PHP Error: Failed to validate JWT: missing aud/azp value in /var/www/html/program/include/rcmail_oauth.php on line 260 (GET /index.php/login/oauth?code=eb631bcb099a4d59ba7cd906aa125a61&state=r5tQZ7phLudP)`

followed by

`errors: <9c5eb26a> PHP Error: OAuth token request failed: Client error: `GET https://sso.DOMAIN.COM/application/o/userinfo/` resulted in a `403 Forbidden` response; roundcube GuzzleHttp/6.5.5 curl/7.74.0 PHP/7.4.27 - [14/Jan/2022:06:57:24 +0000] "GET /application/o/userinfo/ HTTP/1.1" 403  in /var/www/html/program/include/rcmail_oauth.php on line 317 (GET /index.php/login/oauth?code=eb631bcba9934d59ba7cd906aa125a61&state=r5tQZ7phLudP)`

Would you be able to share the property mapping config as well? I am assuming the non-standard dovecot config is the same as you posted before.

**Comment by eglia** [comment]
I just copied the default ``authentik default OAuth Mapping: OpenID 'profile'`` and removed the ``groups`` element:
```
return {
    # Because authentik only saves the user's full name, and has no concept of first and last names,
    # the full name is used as given name.
    # You can override this behaviour in custom mappings, i.e. `request.user.name.split(" ")`
    "name": request.user.name,
    "given_name": request.user.name,
    "family_name": "",
    "preferred_username": request.user.username,
    "nickname": request.user.username,
}
```
Then deselect the default mapping in the oauth2 provider and select this custom mapping.
But I don't think that this is your problem, it seems that the request to the userinfo endpoint fails with error 403 Forbidden. Is there anything in the authentik logs?

**Comment by xpufx** [comment]
Fantastic . Thank you. You might be right about the issue being different but a baseline working conf will help anyway. I already have a separate app/provider setting for the same url with TOTP. That might be the cause of my particular issue. I haven't figured out how to have two different auth methods for the same application yet.

**Comment by xpufx** [comment]
@eglia what is your Roundcube imap_auth_type set to if anything? (same for smtp) It should do xoauth but I can't see that as an option in the defaults file or the wiki.

**Comment by eglia** [comment]
I don't have that set, so it will use the default. Besides the config I posted, everything is at the default values.

**Comment by xpufx** [comment]
Thanks. My roundcube is not even generating anyting in the dovecot logs before it says oauth failed. I have something more fundamental missing. Cheers.

**Comment by xpufx** [comment]
@eglia Hey I got this working. Don't know exactly what it was, i tried a bunch of things in parallel. However once the roundcube login stuff was done Dovecot worked right away thanks to taking care of the property mappings already. Thank you for your help.
